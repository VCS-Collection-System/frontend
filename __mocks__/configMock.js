export const getConfig = (URL, configUrl, config) => {
  if(URL.indexOf(configUrl) > -1 ) {
    const country = URL.substring(configUrl.length);
    const countryConfig = config[country] || null;
    return Promise.resolve({data: countryConfig});
  } else {
    return Promise.reject({data: null});
  }
}