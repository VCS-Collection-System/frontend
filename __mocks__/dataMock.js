export const employeeInfo = {
  "employeeId":"mock1",
  "firstName":"Mock",
  "lastName":"Tester",
  "email":"test@mock.com",
  "agencyCode":"Department of Mock Safety",
  "agencyName":"Department of Mock Safety",
  "divisionCode":"App Mocking",
  "divisionName":null,
  "supervisor":null,
  "dateOfBirth":"2021-10-07",
  "alternateEmail":"mock@email.com",
  "age":0,
  "ageInMonth":0,
  "workforceId":"mock2",
  "hr":false
}

export const attachmentObject = {
  "originalFileName":"blob",
  "size":100,
  "contentType":"image/png",
  "s3BucketName":"mockBucket",
  "s3UUID":"mockUUID"
}

export const vaxHistory = {
  "id":0,
  "employee": employeeInfo,
  "submissionDate":"2021-10-04T15:12:14.875",
  "review":{
     "reviewerEmployeeId":"z2",
     "reviewerNotes":"",
     "reviewDate":"2021-10-05T14:04:29.296",
     "reviewOutcome":"ACCEPTED",
     "rejectReason":""
  },
  "submittedBy":null,
  "autoApproved":false,
  "lastUpdatedTime":null,
  "lastUpdatedBy":null,
  "vaccineBrand":"MODERNA",
  "vaccineAdministrationDate":"2021-10-04",
  "fullVaccinatedFlag":false,
  "vaccineShotNumber":2,
  "documentType":"VaccineDocument",
  "attachment": {}
}

const taskEmployeeObject = {
  "employeeId":"mock1",
  "firstName":"Mock",
  "lastName":"Tester",
  "email":"test@mock.com",
  "agencyCode":"Department of Mock Safety",
  "agencyName":"Department of Mock Safety",
  "divisionCode":"App Mocking",
  "divisionName":null,
  "supervisor":null,
  "dateOfBirth": {
    "year": "2021",
    "monthValue": "12",
    "dayOfMonth": "16"
  },
  "alternateEmail":"mock@email.com",
  "age":0,
  "ageInMonth":0,
  "workforceId":"mock2",
  "hr":false
}

export const mockTask = {
  "task-input-data": {
    "documentList": {
      "com.redhat.vcs.model.VaccineDocumentList": {
        "documents": [ {
          "com.redhat.vcs.model.VaccineCardDocument": {
            "employee": {
              "com.redhat.vcs.model.Employee": taskEmployeeObject
            },
            "vaccineBrand": {
              "com.redhat.vcs.model.VaccineBrand": "MODERNA"
            },
            "attachment": {
              "com.redhat.vcs.model.Attachment": attachmentObject
            },
            "vaccineAdministrationDate": {
              "year": "2021",
              "monthValue": "12",
              "dayOfMonth": "16"
            },
            "lotNumber": "mockLot"
          } }
        }
      }
    }
  }
}
