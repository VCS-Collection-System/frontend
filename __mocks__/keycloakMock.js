module.exports = {
    token: "bogustoken", 
    authenticated: false,
    loadUserProfile: jest.fn( ()=> {return Promise.resolve({
        attributes: {
            workforceid: ["mockId"]
        }
    })} ),
    login: jest.fn(),
    logout: jest.fn(),
    hasRealmRole: jest.fn(),
    profile: {
        firstName: "mockFirst",
        lastName: "mockLast",
        email: "mock@email.com",
        attributes: {
            workforceid: ["mockId"],
            agency: ["USA"]
        }
    }
};
