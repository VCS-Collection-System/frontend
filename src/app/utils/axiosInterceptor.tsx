import axios from "axios";

const axiosInterceptor = axios.create({})

axiosInterceptor.interceptors.request.use(
    config => {
        config.headers['Authorization'] = 'Bearer ' + sessionStorage.getItem("token");
        return config;
    },
    error => {
        Promise.reject(error)
    });
export default axiosInterceptor;