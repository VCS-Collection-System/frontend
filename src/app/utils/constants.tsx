declare global {  
  interface Window {
    REACT_APP_PAM_URL_BASE: string;
    REACT_APP_PAM_CONTAINER_ID: string;
    REACT_APP_KEYCLOAK_COUNTRY: string;
    REACT_APP_KEYCLOAK_EMAIL: string;
    REACT_APP_KEYCLOAK_FIRST_NAME: string;
    REACT_APP_KEYCLOAK_ID: string;
    REACT_APP_KEYCLOAK_LAST_NAME: string;
    REACT_APP_MAINTENANCE_BANNER: boolean;
    REACT_APP_MAINTENANCE_BANNER_TEXT: string;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    REACT_APP_FEATURE_FLAGS: Array<Record<string, any>>;
    REACT_APP_MAX_DOSES: Record<string, number>;
    REACT_APP_CONSENT_LABEL: string;
    REACT_APP_ALTERNATE_CONSENT_LABEL: string;
    REACT_APP_BLOCKED_COUNTRIES: Array<string>;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    REACT_APP_GLOBAL_CONFIG: Record<string, any>;
  }
}

// PAM urls
export const PAM_BASE_URL = window.REACT_APP_PAM_URL_BASE; 
export const SUBMIT_VAX_URL = PAM_BASE_URL + '/attestation/vax';
export const SUBMIT_TEST_URL = PAM_BASE_URL + '/attestation/covid-test-result';
export const PAM_TASK_URL = PAM_BASE_URL + '/tasks/';
export const RESERVED_TASK_URL = PAM_TASK_URL + 'owners?status=Reserved&user=';
export const INPROGRESS_TASK_URL = PAM_TASK_URL + 'owners?status=InProgress&user=';
export const ALL_TASK_URL = PAM_TASK_URL + 'owners?status=Ready&pageSize=200';
export const PRIORITY_URL = PAM_BASE_URL + '/query/priority-results';
export const PRIORITY_CONFIRM_URL = PAM_BASE_URL + '/query/priority-result/';
export const ATTACHMENT_URL = PAM_BASE_URL + '/attestation/attachment';
export const EMPLOYEE_URL = PAM_BASE_URL + '/query/employee/';
export const EMPLOYEE_NAME_URL = PAM_BASE_URL + '/query/employee-name/';
export const EMPLOYEE_VAX_HISTORY_URL = PAM_BASE_URL + '/query/vax-document/';
export const EMPLOYEE_TEST_HISTORY_URL = PAM_BASE_URL + '/query/test-result-document/';
export const VAX_TASK_DOCUMENT_URL = PAM_BASE_URL + '/query/task/document/';
export const DELETE_VAX_URL = PAM_BASE_URL + '/attestation/delete-vax-record/';
export const CONFIG_URL = PAM_BASE_URL + '/query/country-configs/';
export const SAVE_CONFIG_URL = PAM_BASE_URL + '/attestation/country-configs';

// List of countries allow to use this app
export const BLOCKED_COUNTRIES = window.REACT_APP_BLOCKED_COUNTRIES || []

export const KONAMI_CODE = [ "ArrowUp", "ArrowUp", "ArrowDown", "ArrowDown", "ArrowLeft", "ArrowRight", "ArrowLeft", "ArrowRight", "KeyB", "KeyA"];

// Maintenance Banner
export const MAINTENANCE_BANNER = window.REACT_APP_MAINTENANCE_BANNER || false
export const MAINTENANCE_BANNER_TEXT = window.REACT_APP_MAINTENANCE_BANNER_TEXT || ""

// Keycloak claims
export const KEYCLOAK_COUNTRY = window.REACT_APP_KEYCLOAK_COUNTRY
export const KEYCLOAK_EMAIL = window.REACT_APP_KEYCLOAK_EMAIL
export const KEYCLOAK_FIRST_NAME = window.REACT_APP_KEYCLOAK_FIRST_NAME
export const KEYCLOAK_ID = window.REACT_APP_KEYCLOAK_ID
export const KEYCLOAK_LAST_NAME = window.REACT_APP_KEYCLOAK_LAST_NAME
export const FEATURE_FLAGS = window.REACT_APP_FEATURE_FLAGS || []

// Maximum doses allowed by vaccine manufacturer
export const MAX_DOSES = window.REACT_APP_MAX_DOSES || {}

export const CONSENT_LABEL = window.REACT_APP_CONSENT_LABEL
export const ALTERNATE_CONSENT_LABEL = window.REACT_APP_ALTERNATE_CONSENT_LABEL

// Global country config
export const GLOBAL_CONFIG = window.REACT_APP_GLOBAL_CONFIG || {};

