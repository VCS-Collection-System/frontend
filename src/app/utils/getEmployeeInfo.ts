import {
  KEYCLOAK_FIRST_NAME, 
  KEYCLOAK_LAST_NAME, 
  KEYCLOAK_EMAIL,
  KEYCLOAK_COUNTRY,
  KEYCLOAK_ID
} from "@app/utils/constants"; 

export function getEmployeeInfo(keycloak) {
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  const jp = require('jsonpath');
  const profile = keycloak.profile;
  const employeeFirstName = jp.query(profile, KEYCLOAK_FIRST_NAME)[0];
  const employeeLastName = jp.query(profile, KEYCLOAK_LAST_NAME)[0];
  const employeeEmail = jp.query(profile, KEYCLOAK_EMAIL)[0];
  const employeeCountry = jp.query(profile, KEYCLOAK_COUNTRY)[0];
  const employeeID = jp.query(profile, KEYCLOAK_ID)[0];

  const isApprover = () => {
    if(employeeCountry && 
      (keycloak.hasRealmRole("vcs-superapprover") || keycloak.hasRealmRole("vcs-approver-" + employeeCountry.toLowerCase()))){
        return true;
    }

    return false;
  }

  const isSuperApprover = () => {
    return keycloak.hasRealmRole("vcs-superapprover");
  }

  return {
    firstName: employeeFirstName, 
    lastName: employeeLastName,
    email: employeeEmail,
    country: employeeCountry,
    id: employeeID,
    isApprover: isApprover(),
    isSuperApprover: isSuperApprover()
  }
}
