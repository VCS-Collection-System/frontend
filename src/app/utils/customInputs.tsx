import { useEffect, useState } from "react";
import { KONAMI_CODE } from './constants'

const useInputEvent = () => {
  const [key, setKey] = useState(null);
  useEffect(() => {
    const keyDownHandler = ({ code }) => setKey(code);
    const keyUpHandler = () => setKey(null);
    window.addEventListener('keydown', keyDownHandler);
    window.addEventListener('keyup', keyUpHandler);
    return () => {
      window.removeEventListener("keydown", keyDownHandler);
      window.removeEventListener("keyup", keyUpHandler)
    }
  }, []);
  return key;
}

const useSecretCode = (secretCode) => {
  const [count, setCount] = useState(0);
  const [success, setSuccess] = useState(false);
  const key = useInputEvent();
  
  useEffect(() => {
    if (key == null) return;
    if (key !== secretCode[count]) {
      setCount(0);
      return;
    }
    
    setCount((state) => state + 1);
    if (count + 1 === secretCode.length) {
      setSuccess(!success);
      setCount(0);
    }
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [key]); 
  
  return success;
};

export const useDarkMode = () => {
  return useSecretCode(KONAMI_CODE);
};
