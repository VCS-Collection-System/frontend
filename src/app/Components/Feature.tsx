import React, { ReactElement, useEffect, useState } from 'react';
import { useKeycloak } from "@react-keycloak/web";
import { getEmployeeInfo } from "@app/utils/getEmployeeInfo";
import authaxios from "@app/utils/axiosInterceptor";
import { CONFIG_URL, FEATURE_FLAGS, GLOBAL_CONFIG } from "@app/utils/constants";

interface FeatureProps {
  name: string;
  employeeCountry?: string;
  renderOff?: () => ReactElement;
  children?: React.ReactNode
}

const getFeatureFlag = (name) => {
  const feature = FEATURE_FLAGS.find(featureFlag => featureFlag.name === name);
  return feature ? feature : {}; 
}

const isFeatureFlag = (name) => {
  return FEATURE_FLAGS.find(featureFlag => featureFlag.name === name);
}

const getConfig = async (country) => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  let defaultConfig: any = null;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  let countryConfig: any = null;
  const sessionDefault = sessionStorage.getItem("globalconfig-default");
  const sessionCountry = sessionStorage.getItem("globalconfig-" + country);

  if(sessionDefault && sessionDefault !== "") {
    defaultConfig = JSON.parse(sessionDefault);
  } else {
    try {
      await authaxios.get(CONFIG_URL + "default").then(config => {
        defaultConfig = config.data;
        sessionStorage.setItem("globalconfig-default", JSON.stringify(config.data));
      });
    } catch {
      defaultConfig = GLOBAL_CONFIG["default"];
    }
  }

  if(sessionCountry && sessionCountry !== "") {
    countryConfig = JSON.parse(sessionCountry);
  } else {
    try {
      // all db gets/persists must use lowercase country
      await authaxios.get(CONFIG_URL + country.toLowerCase()).then(config => {
        countryConfig = config.data;
        sessionStorage.setItem("globalconfig-" + country, JSON.stringify(config.data));
      });
    } catch {
      countryConfig = defaultConfig;
    }
  }

  return [defaultConfig, countryConfig];
}

const showForCountry = async (name, keycloak, employeeCountry) => {
  let userCountry = "default";
  if(keycloak.profile) {
    // get user country if logged in
    const employee = getEmployeeInfo(keycloak);
    if(employeeCountry && employeeCountry !== "" && employee && employee.isApprover) {
      userCountry = employeeCountry;
    } else if(employee && employee.country) {
      userCountry = employee.country;
    }
  }

  const [defaultConfig, countryConfig] = await getConfig(userCountry);

  // get default config for feature
  let featureConfig = false;
  if(defaultConfig[name] != null) {
    featureConfig = defaultConfig[name];
  } else {
    // if not defined in default config, return false
    return featureConfig;
  }

  // get country specific config
  if(countryConfig && countryConfig[name] != null) {
    // use country-specific feature override
    featureConfig = countryConfig[name];
  }

  return featureConfig;
}

const Feature = ({name, employeeCountry, renderOff, children}: FeatureProps) => {
  const { keycloak } = useKeycloak();
  const [visible, setVisible] = useState(false);

  useEffect(() => {
    showFeature(name, keycloak, employeeCountry).then(show => {
      setVisible(show);
    })
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  if(visible) {
    return (<>{children}</>)
  } else {
    return (renderOff ? renderOff() : null);
  }
}; 

export const showFeature = async (name, keycloak, employeeCountry = "") => {
  const show = await showForCountry(name, keycloak, employeeCountry);
  return (isFeatureFlag(name) ? getFeatureFlag(name).active && show : show);
}

export default Feature;
