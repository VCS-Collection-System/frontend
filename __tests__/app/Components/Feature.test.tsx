import React from "react"
import Feature, { showFeature } from "@app/Components/Feature"
import { render, screen } from '@testing-library/react'
import mockKeycloak from "../../../__mocks__/keycloakMock"
import { getConfig } from "../../../__mocks__/configMock"
import { CONFIG_URL } from "@app/utils/constants"

jest.mock("@react-keycloak/web", () => ({
  useKeycloak: jest.fn(() => ({ keycloak: mockKeycloak })),
}))

jest.mock("@app/utils/constants", () => ({
  CONSENT_COUNTRIES: ["AUS"],
  FEATURE_FLAGS: [
    {
      "name": "fakeFeature", 
      "active": true
    },
    { 
      "name": "hrdashboard",
      "active": true
    }, 
    { 
      "name": "inactive",
      "active": false
    }, 
    {
      "name": "covidTest",
      "active": true
    },
    {
      "name": "consentCheckbox",
      "active": true
    },
    {
      "name": "customInput",
      "active": true
    }
  ],
  KEYCLOAK_COUNTRY: "attributes.agency[0]",
  KEYCLOAK_EMAIL: "email",
  KEYCLOAK_FIRST_NAME: "firstName",
  KEYCLOAK_LAST_NAME: "lastName",
  KEYCLOAK_ID: "attributes.workforceid[0]",
  CONFIG_URL: "config/"
}));

const config = {"default": {
  "allowPdf": false,
  "dateOfBirth": true,
  "consentCheckbox": false,
  "covidTest": false,
  "customInput": true,
  "hrdashboard": true,
  "hrsearch": true,
  "cdc": false,
  "divoc": false,
  "greenPass": false,
  "recovery": false,
  "fakeFeature": true
}};

jest.mock('@app/utils/axiosInterceptor', () => ({
  get: (URL: string) => getConfig(URL, CONFIG_URL, config)
}))

describe('<Feature />', ()=> {
  it('should be defined', () => {
    expect(Feature).toBeDefined()
  })
  it("should match existing snapshot", () => {
    expect(render(<Feature name="fakeFeature"><div /></Feature>).asFragment()).toMatchSnapshot()
  })

  it("should render child component when set in feature flag", async () => {
    render(
      <Feature name="fakeFeature">
        <div data-testid="mockComponent"/>
      </Feature> 
    )
    await new Promise(process.nextTick);
    expect(screen.getByTestId('mockComponent')).toBeTruthy();
  })

  it("should not render child component when not in feature flag", async () => {
    render(
      <Feature name="fakeFeature2">
        <div data-testid="mockComponent"/>
      </Feature>
    )
    await new Promise(process.nextTick);
    expect(screen.queryByTestId('mockComponent')).toBeFalsy();
  })

  it("should not render child component when active is false", async () => {
    render(
      <Feature name="inactive">
        <div data-testid="mockComponent"/>
      </Feature> 
    )
    await new Promise(process.nextTick);
    expect(screen.queryByTestId('mockComponent')).toBeFalsy();
  })

  it("should render off when not in feature", async () => {
    render(
      <Feature 
        name="notfeature"
        renderOff={() => <div data-testid="mockComponentOff"/> }
      >
        <div data-testid="mockComponentOn"/>
      </Feature> 
    )
    await new Promise(process.nextTick);
    expect(screen.getByTestId('mockComponentOff')).toBeTruthy();
  })

  it("should return whether to show feature", async () => {
    showFeature("fakeFeature", mockKeycloak).then(actualObject => {
      expect(actualObject).toEqual(true); 
    });

    showFeature("inactive", mockKeycloak).then(actualObjectTwo => {
      expect(actualObjectTwo).toEqual(false); 
    });
  })

  it("should return false when not defined", async () => {
    showFeature("notFeature", mockKeycloak).then(actualObject => {
      expect(actualObject).toEqual(false);
    })
  });
})