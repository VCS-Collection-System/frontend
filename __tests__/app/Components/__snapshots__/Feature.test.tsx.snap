// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Feature /> should match existing snapshot 1`] = `<DocumentFragment />`;
